Feature/Extended coverage for Vehicle Data tests

dboltovskyi opened the pull request:
This is base PR for test scripts for Vehicle Data proposals:
 - [SDL-0236] Update mismatch in TireStatus structure
 - [[SDL 0255] Feature/Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2489)
 - [[SDL 0262] Feature/New vehicle data SeatOccupancy](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2490)
 - [[SDL 0269] Feature/New vehicle data ClimateData](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2491)

This PR is **[ready]** for review.

### Summary
Extended test coverage for Vehicle Data tests

### ATF version
[7.1.0_version_bump](https://github.com/smartdevicelink/sdl_atf/pull/217)

### Changelog

- Introduced new modules
    - `APIHelper` - set of utility functions to work with both APIs: MOBILE and HMI
    - `APITestDataGenerator` - generate values for parameters based on existing API restrictions and desired value type

- Introduced the following test types to verify `GetVD` and `OnVD` RPCs:
    - `VALID_RANDOM_ALL` - positive cases for VD parameters where all possible sub-parameters of hierarchy are defined with valid random values,
    - `VALID_RANDOM_SUB` - positive cases for struct VD parameters and sub-parameters where only one sub-parameter of hierarchy is defined with valid random value (mandatory also included),
    - `LOWER_IN_BOUND/UPPER_IN_BOUND` - positive cases for VD parameters and sub-parameters where only one sub-parameter of hierarchy is defined with min/max valid value (mandatory also included),
    - `LOWER_OUT_OF_BOUND/UPPER_OUT_OF_BOUND` - negative cases for VD parameters and sub-parameters where only one sub-parameter of hierarchy is defined with nearest invalid min/max value,
    - `INVALID_TYPE` - negative cases for VD parameters and sub-parameters with invalid type value defined for one of them,
    - `ENUM_ITEMS/ BOOL_ITEMS` - positive cases for enum/boolean VD parameters and sub-parameters with all possible values defined (all items for enum and true/false for boolean),
    - `MANDATORY_ONLY` - positive cases for struct VD parameters and sub-parameters where only mandatory sub-parameters are defined with valid random values,
    - `MANDATORY_MISSING` - negative cases for struct VD parameters and sub-parameters where only mandatory sub-parameters are defined and one of them is missing,
    - `PARAM_VERSION` - positive/negative cases for VD parameters with version defined
    
- Deleted existing VD scripts which are covered by a new tests
- Updated existing scripts
- Updated `vehicle_data.txt` test set
- Added possibility to restrict vehicle data parameters to be tested by defining `VD_PARAMS` environment variable

##### Bug Fixes
- Fixed issue of handling API data types without `Common.` prefix
- Improved generation of random values

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

KhrystynaDubovyk commented:
@atiwari9 please review this PR

atiwari9 commented on user_modules/api/APITestDataGenerator.lua:
```diff
@@ -0,0 +1,308 @@
+local function getStringValue(pTypeData)
+  local length
+  if pTypeData.valueType == m.valueType.LOWER_IN_BOUND then
+    length = pTypeData.minlength
```
Is this minlength taken from API for the parameter in question?

dboltovskyi replied:
Correct.

atiwari9 commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +502,672 @@ function m.setAppVersion(pParamVersion, pOperator)
+    mobile = {
+      name = rpc,
+      request = request,
+      response = { success = false, resultCode = "GENERIC_ERROR" }
```
Shouldn't the response here be `INVALID_DATA`

dboltovskyi replied:
SDL usually responds with `INVALID_DATA` in case some request from Mobile has incorrect data.
However in our case we have incorrect data sent from HMI. So SDL should reply `GENERIC_ERROR`.

Please find a requirement: 
> In case
HMI sends invalid response by any reason that SDL must transfer to mobile app
SDL must:
log the error internally
respond GENERIC_ERROR (success:false, info: "Invalid message received from vehicle") to mobile app

atiwari9 commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -46,14 +43,21 @@ do
+  [m.rpc.on] = "VehicleInfo.OnVehicleData"
+}
+
 m.vd = {
```
Is it possible to run tests for a specific vehicle data items? From looks of it, these changes generalize the VD testing for all params and there is no way to run tests for a singular or a set of VD items.

dboltovskyi replied:
Currently this can be achieved by commenting values in `m.vd` table.

atiwari9 replied:
Ok, but that is not really clean approach. Is it possible to set some kind of environment variable or configuration for that affect. As a test runner, i'd not like to modify the lua scripts to target specific test runs.

dboltovskyi replied:
That's a good idea. I'm going to push an update regarding this in a few hours.

atiwari9 commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -7,20 +7,17 @@ local runner = require('user_modules/script_runner')
 --[[ General configuration parameters ]]
 runner.testSettings.isSelfIncluded = false
 config.defaultProtocolVersion = 2
+config.zeroOccurrenceTimeout = 1000
```
What does zero occurrence mean? What is this used for?

dboltovskyi replied:
This parameter defines a timeout for SDL to wait for an events that should not occurs.
For example in ATF script we're expecting `OnVehicleData` should not be transfered by SDL to Mobile.
In this case we register expectation with `Times(0)`:
```
common.getMobileSession():ExpectNotification("OnVehicleData"):Times(0)
```
Having this ATF will wait until timeout defined by `zeroOccurrenceTimeout` parameter is expired and then finish test step.

Decreasing default value from `2000` to `1000` ms allows us to run VD test scripts faster.

atiwari9 commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -100,185 +104,100 @@ m.app = {
+m.isMandatory = {
+  YES = true,
+  NO = false,
+  ALL = 3
```
Does ALL mean both cases?

dboltovskyi replied:
Correct.

atiwari9 commented:
@dboltovskyi  - There are significant changes to testing methodology, these changes warrant full regression tests for all VD items including the tests which are removed. I am skeptical of the scope of these changes with-in new VD related proposals, we may need to check with PMs on if this needs to be part of your other ATF related change set. 

I'd run the full test suite myself. Meanwhile i have left some comments for the changes.

dboltovskyi commented:
@atiwari9 Regarding your comment [2488#issuecomment-735976762](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#issuecomment-735976762) please let me add some clarification about the changes in this PR.

Use of APIs to generate tests and test cases had been introduced in a previous release (`7.0`) (please see https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2439)
There was a request from PMs to have a generic approach for testing all VD parameters.

Current PR is an evolution of existing approach which also adds systematization in common modules and tests.

Despite the fact table with test types has been introduced in the current PR, the following checks have been implemented before (in a previous release):
 - `VALID_RANDOM_ALL`, 
 - `INVALID_TYPE`, 
 - `MANDATORY_ONLY`, 
 - `MANDATORY_MISSING`,
 - `PARAM_VERSION`

 Current PR provides the following changes (comparing to existing approach):
 - Introduced modules to work with APIs (`APIHelper` and `APITestDataGenerator`). Previously these were a part of `common` module
 - Introduced a few tables in `common` module (e.g. `m.testType`)
 - Added new functions in `common` module to generate test cases (e.g. `m.createTestCases`)
 - Introduced possibility to recursively check all sub-parameters
 - Added new test types and corresponding scripts:
   - `VALID_RANDOM_SUB`
   - `LOWER_IN_BOUND/UPPER_IN_BOUND`
   - `LOWER_OUT_OF_BOUND/UPPER_OUT_OF_BOUND`,
   - `ENUM_ITEMS/ BOOL_ITEMS`
 - Updated existing scripts to utilize new functions
 - Deleted scripts which are covered by a new approach

I agree there are quite a lot of changes. Though having them we will significantly improve existing approach and I would say "finalize" it. I hope PMs wouldn't mind to accept the changes.

atiwari9 commented:
@dboltovskyi  Also please add new testscripts to vehicle data test sets.

atiwari9 commented:
@dboltovskyi  - I ran the vehicle_data test set only for tirePressure and got 5 Failed cases, is this expected?

TOTAL: 38
PASSED: 29 
FAILED: 5 
ABORTED: 4 

Logs: 
[2020-12-01_15-10-58.736.zip](https://github.com/smartdevicelink/sdl_atf_test_scripts/files/5625358/2020-12-01_15-10-58.736.zip)

```
Test target: test_sets/vehicle_data.txt
----------------------------------------------------------------------------------------------------
01:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/001_GetVD_Success_flow.lua	
02:	ABORTED	./test_scripts/API/VehicleData/GetVehicleData/002_GetVD_disallowed.lua	
03:	FAILED	./test_scripts/API/VehicleData/GetVehicleData/003_GetVD_disallowed_after_PTU.lua	
04:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/004_GetVD_HMI_responds_with_invalid_data.lua	
05:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/005_GetVD_App_sends_invalid_request.lua	
06:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/006_GetVD_mandatory_parameters.lua	
07:	FAILED	./test_scripts/API/VehicleData/GetVehicleData/007_GetVD_app_version_is_greater_than_parameter_version.lua	
08:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/008_GetVD_app_version_is_less_than_parameter_version.lua	
09:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/009_GetVD_min_max_boundary_values.lua	
10:	PASSED	./test_scripts/API/VehicleData/GetVehicleData/010_GetVD_enum_and_bool_values.lua	
11:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/001_OnVD_Success.lua	
12:	ABORTED	./test_scripts/API/VehicleData/OnVehicleData/002_OnVD_disallowed.lua	
13:	FAILED	./test_scripts/API/VehicleData/OnVehicleData/003_OnVD_disallowed_after_PTU.lua	
14:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/004_OnVD_with_invalid_data.lua	
15:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/005_OnVD_App_is_not_subscribed.lua	
16:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/006_OnVD_mandatory_parameters.lua	
17:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/007_OnVD_min_max_boundary_values.lua	
18:	PASSED	./test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua	
19:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/001_SubscribeVD_Success.lua	
20:	ABORTED	./test_scripts/API/VehicleData/SubscribeVehicleData/002_SubscribeVD_disallowed.lua	
21:	FAILED	./test_scripts/API/VehicleData/SubscribeVehicleData/003_SubscribeVD_disallowed_after_PTU.lua	
22:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/004_SubscribeVD_HMI_responds_with_invalid_data.lua	
23:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/005_SubscribeVD_App_sends_invalid_request.lua	
24:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/006_SubscribeVD_2nd_request_IGNORED.lua	
25:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/007_SubscribeVD_for_2_apps.lua	
26:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/008_SubscribeVD_HMI_responds_with_not_success_codes.lua	
27:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/009_SubscribeVD_Resumption_of_subscription_Unexpected_Disconnect.lua	
28:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/010_SubscribeVD_Resumption_of_subscription_Ignition_Cycle.lua	
29:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/011_SubscribeVD_Resumption_of_subscription_for_2_apps_Unexpected_Disconnect.lua	
30:	PASSED	./test_scripts/API/VehicleData/SubscribeVehicleData/012_SubscribeVD_Resumption_of_subscription_for_2_apps_Ignition_Cycle.lua	
31:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/001_UnsubscribeVD_Success.lua	
32:	ABORTED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/002_UnsubscribeVD_disallowed.lua	
33:	FAILED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/003_UnsubscribeVD_disallowed_after_PTU.lua	
34:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/004_UnsubscribeVD_HMI_responds_with_invalid_data.lua	
35:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/005_UnsubscribeVD_App_sends_invalid_request.lua	
36:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/006_UnsubscribeVD_2nd_request_IGNORED.lua	
37:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/007_UnsubscribeVD_for_2_apps.lua	
38:	PASSED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/008_UnsubscribeVD_HMI_responds_with_not_success_codes.lua	
```

dboltovskyi commented:
@atiwari9 Please notice approach with commenting of parameters in `m.vd` table is not working for all scripts.
I'm going to push an update with possibility to define tested VD parameters in environment variable which will work for all VD scripts.

dboltovskyi commented:
@atiwari9 Please find update in 27af7bf 

Now it's possible to define VD parameters for all test scripts in `VD_PARAMS` environment variable.
It's allowed to define one or many parameters split by coma. E.g: `export VD_PARAMS=gps` or `export VD_PARAMS=gps,speed`.
If `VD_PARAMS` is not defined tests will be run against all possible VD parameters.

One limitation - this approach doesn't work in ATF parallel mode by default since it requires changes in ATF.

However if you would like to do so, you just need to add the following line after https://github.com/smartdevicelink/sdl_atf/blob/6f600b7c8e926e569c9e291973817b21c850177e/atf_parallels/loop.sh#L29        
```
-e VD_PARAMS=$VD_PARAMS \
```
This will transfer `VD_PARAMS` environment variable into `atf_worker` docker container.

atiwari9 commented:
@dboltovskyi  - Thanks, this seems to be working so i am running another batch of tests now. Is there a place to document this capability of running selectively vehicle data item tests?

dboltovskyi commented:
@atiwari9 For now it's "undocumented" feature.
There only place where it's mentioned is a description for `restrictAvailableVDParams()` function:
https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/27af7bf38fc552476fd05276bac7416f39064523/test_scripts/API/VehicleData/common.lua#L175

Additionally we can do the following:
1. Update current PR in Scripts with description of this possibility
2. Create a new PR in ATF repo with update for the README (please see `Advanced options` section ) and also with update for parallel mode

What do you think would be the best?

atiwari9 commented:
@dtrunov - I think option 2 is better, as a user may refer to ReadMe to understand the usage rather than open each and every script to find out.

BTW, with commit 27af7bf , All tests are Passed for individual `tirePressure` as well as batch for all VD.

Logs: 

[All_Tests_Pass.7z.changeExtTo7z.log](https://github.com/smartdevicelink/sdl_atf_test_scripts/files/5631633/All_Tests_Pass.7z.changeExtTo7z.log)
[TirePressure_AllTests_Pass.zip](https://github.com/smartdevicelink/sdl_atf_test_scripts/files/5631806/TirePressure_AllTests_Pass.zip)

dboltovskyi commented:
@atiwari9 Agree. I will do both: 1) and 2) and post a comment when it's ready

dboltovskyi commented:
@atiwari9 Please notice:
1. Description of the current PR is updated with note regarding `VD_PARAMS` environment variable
2. New PR in ATF is created with appropriate updates: https://github.com/smartdevicelink/sdl_atf/pull/218

atiwari9 reviewed (approved):
Approved

LitvinenkoIra commented:
@JackLivio This PR is ready for Livio review. Thank you!

jacobkeeler commented on test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua:
```diff
@@ -0,0 +1,38 @@
+common.getTestsForOnVD(testTypes)
+
+common.Title("Postconditions")
+common.Step("Stop SDL", common.postconditions)
```
```suggestion
common.runner.Step("Stop SDL", common.postconditions)
```
I think it would be better to use the runner as a component of the common module, rather than just taking functions directly from the runner.
Calling script runner functions via the common module is confusing

dboltovskyi replied:
Done in 53d0806

jacobkeeler commented on user_modules/api/APIHelper.lua:
```diff
@@ -0,0 +1,195 @@
+--]]
+function m.getGraph(pAPIType, pEventType, pFuncName)
+
+  local function getGraph(pParams, pGraph, pParentId)
```
From what I can tell this is just flattening out the map of all the parameters into a single array. 

Is there any reason for this approach? It seems to just make all of the logic more complex and confusing as far as I can tell.

dboltovskyi replied:
The main benefit of this approach is to have possibility to check not only `root` parameters (as before), but also other sub-parameters on any level of hierarchy.
E.g. in case of root VD parameter `bodyInformation` it's possible to check all children sub-parameters:
```
- bodyInformation
  - roofStatuses
    - location
      - rowspan
```

jacobkeeler replied:
Seems to me that a recursive approach would be easier to write and less confusing. I don't really see any benefit to having all of these parameters on the same level.

dboltovskyi replied:
I apologize if I was not describe the benefits of the approach in a clear way. Let me add some details.
For example we have the following structure:
```
(1) - bodyInformation
(2)   - roofStatuses
(3)     - location
(4)       - rowspan
```

Old approach doesn't allow us to check any parameter except of the root one (`bodyInformation`). All children parameters (`roofStatuses`, `location`, `rowspan`) would be defined with some valid random values and that's it.

Within new approach each sub-parameter of hierarchy is verified against API restrictions.

E.g. we're able to check `rowspan` parameter which is on the lowest level of hierarchy against `maxvalue="100"` API restriction.

In order to do so we need to build the minimal version of hierarchy of parameters up to the root.
This hierarchy tree should include mandatory parameters on each levels and sub-levels in order to have a valid scenario.
That's why we need all these functions: `getParents`, `getMandatoryNeighbors`, `getMandatoryChildren`
All the parameters in the tree should have valid values defined.

Then we can test `rowspan` parameter against any other restriction defined in API.

New approach iterates through all the levels of hierarchy and builds all possible cases depends on rules defined.
For example it's possible to build the cases for Arrays only, for Strings, for Mandatory etc.

jacobkeeler replied:
But this isn't the only way to check subparameters, correct? You should still be able to do so by moving recursively through the parameter map, and I'd argue doing so would be simpler and more intuitive. Flattening out the map into an array seems like an unnecessary step.

Is there any part of the logic that wouldn't be possible just using a map of all the parameters?

dboltovskyi replied:
I would say flattened map (graph) is main concept of the new approach.
This artificial simplified representation of API structures is much easier to use along with iterative functions implemented in:
 - `APITestDataGenerator` module to build a tree of parameters and their values.
 - `VD common` module to generate test cases.

Taking into account complexity of the task (testing of all VD parameters and sub-parameters by different test types) it would take considerable time to re-work and re-test a new approach with recursion and without flattened map.

jacobkeeler replied:
After looking through the code again, I can see there are a few places where it would be fairly difficult to refactor (mainly in logic tied to the `getTestCases` function). 

It does seem like the code isn't quite as confusing once you fully understand the structure of the graph, but I think the issue is that the structure is not terribly obvious to anyone reading the code, so I think this approach would be more acceptable if the code were commented better. Specifically, I would suggest adding clarifying comments throughout the `getTestCases` function, as well as a more detailed explanation of the general structure of the graph in the description for the `getGraph` function (such as the fact that it includes all subparameters as well as root-level parameters).

dboltovskyi replied:
Please find update in afb7faf

jacobkeeler commented on user_modules/api/APITestDataGenerator.lua:
```diff
@@ -0,0 +1,308 @@
+  local length
+  if pTypeData.valueType == m.valueType.LOWER_IN_BOUND then
+    length = pTypeData.minlength
+    if not length or length == 0 then length = ah.dataType.STRING.min end
```
Is there a reason that `length == 0` is not allowed here? Empty strings are allowed by the API

dboltovskyi replied:
Even though API allows to have 0-length string SDL will cut off such parameter from request/response.
E.g. in case of `cloudAppVehicleID` which is String type if HMI provides `cloudAppVehicleID=""` SDL will cut off this parameter from response. And then if there are no other VD parameters this response will be treated as invalid.

```
HMI->SDL [08-01-2021 00:38:31,782]
{
    "id": 137,
    "jsonrpc": "2.0",
    "result": {
        "cloudAppVehicleID": "",
        "code": 0,
        "method": "VehicleInfo.GetVehicleData"
    }
}
 
SDL->MOB [08-01-2021 00:38:31,814] 
[rpcFunction: GetVehicleData, sessionId: 1, version: 2, frameType: 1, encryption: false, serviceType: 7, frameInfo: 0, messageId: 115, binaryDataSize: 0] : 
{
    "info": "Invalid message received from vehicle",
    "resultCode": "GENERIC_ERROR",
    "success": false
}


```

dboltovskyi replied:
Same is applicable for empty arrays, e.g. `{}`

jacobkeeler replied:
`cloudAppVehicleID` has a minlength of 1 (the property defaults to 1 if omitted), so it make sense that Core would mark it as invalid. 

It doesn't look like there is any vehicle data value which allows an empty string as of now, but that doesn't mean that it couldn't happen in the future.

dboltovskyi replied:
Interesting ...
I made an experimental SDL build from https://github.com/dboltovskyi/sdl_core/tree/tmp/vd_min_len with `minlength="0"` defined for `cloudAppVehicleID` and got the same result. I.e. empty-string is treated by SDL as invalid value.

dboltovskyi replied:
Interesting ...
I made an experimental SDL build from https://github.com/dboltovskyi/sdl_core/tree/tmp/vd_min_len branch.
It has `minlength="0"` defined for `cloudAppVehicleID` for `GetVD` and `OnVD` RPCs.
Result is different comparing to develop.
Now empty string is treated by SDL as valid value.
Though empty arrays are still invalid.
Please find update in 08a37fe  which returns back exclusion for String parameters with minlength="0"

jacobkeeler replied:
I think that this line would still need to be changed as well
```
if not length then length = api.dataType.STRING.min end
```
To account for the fact that `STRING.min=1`
```
  STRING = { type = "String", min = 1, max = 100 },
```

jacobkeeler replied:
Also, what do you mean the same happens with arrays? The same default applies, `minsize="1"` if omitted. If you set `minsize="0"`, then it will accept an empty array (although as demonstrated several places in this PR, you need to specify an empty array using `json.EMPTY_ARRAY`, since lua cannot differentiate an empty array from an empty map).

dboltovskyi replied:
1\) As for the `String` parameters:

There are default restrictions in `APIHelper.dataType` table.
For example if some `String` parameter have no restrictions defined (min or max) then default restrictions would take place.
Since default min for String is `1` it means the following:
   - length for `LOWER_IN_BOUND` would be be `1` (api.dataType.STRING.min)
   - length for `LOWER_OUT_OF_BOUND` would be `0` (api.dataType.STRING.min-1)

Another scenario if some `String` parameter have `minLength=0` then:
 - length for `LOWER_IN_BOUND` would be `0`
 - there would be no `LOWER_OUT_OF_BOUND` case generated

2\) As for the `Array` parameters:

Good catch for `json.EMPTY_ARRAY` const.

Please find update for points 1) and 2) in 33ecefb

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+    end
+    return nil
+  end
+  local paramId = getParamId(graph, paramName)
```
Function `getParamId()` allows to find `id` in the graph of the root VD parameter by it's `name`.
Then this `id` is used to get the branch of the graph which corresponds to only single VD parameter.

jacobkeeler replied:
It still seems to me that these two would be identical, based on the logic (both of these would be the `key` for the parameter).

Can you give me an example where `paramId` is different than `paramName`?

dboltovskyi replied:
Seems I got the question.

Initialy in a previous approach name of the parameter was used to identify a parameter.
It was Ok since only `root` level was verified.
However in a new approach it was found that can exist 2 parameters with the same `name` in a different branches of hierarchy.
E.g. 

1) `pressure`:
  - tirePressure
    - leftFront 
      - pressure 
    - rightFront
      - pressure

2) `col`
 - bodyInformation
   - doorStatuses
     - location
       - col
    - gateStatuses
      - location
        - col

In order to have unique identifier for each parameter (and sub-parameter) it was decided to move from `name` to `id`

jacobkeeler replied:
Alright, I can see the case where a different identifier would be necessary, but I still do not see how this logic accomplishes this:
```
local function getParamId(pGraph, pName)
    for k, v in pairs(pGraph) do
      if v.parentId == nil and v.name == pName then return k end
    end
    return nil
  end
```
all this seems to do is return the name of the param (if it exists at the root level). Is `k` in this case the full parameter path (ex. `tirePressure.leftFront.pressure`)? 

It seems to me that it is only the name (ex. `pressure`, which is already contained in `paramName`)

dboltovskyi replied:
I guess I need to explain what is the `graph` object.
Graph object in these tests is a table which represents hierarchy of all VD parameters (incl. sub-parameters).

`key` in this table is unique identifier of the VD parameter, this is an integer type value.
`value` in this table is a set of elements. Most important ones are: 
 - `parentId` - identifier of a parent (or nil for the root elements), this is also integer value.
 - `name` - name of the parameter
 - `<restrictions>`, such as `type`, `array`, `mandatory`, `since` etc. copied from API

`graph` object can be easily traversed on any level of hierarchy and is used to build tree of parameters and it's values.

Having that `getParamId()` function returns `Id` of the parameter (`Integer`) by it's name (`String`)

jacobkeeler replied:
Alright, I believe I understand now. However, this does somewhat tie into my point that this "graph" approach isn't particularly intuitive or clearly described. Using this approach just seems to complicate the logic as far as I can tell.

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -339,64 +299,27 @@ function m.getHashId(pAppId)
-
---[[ @getMandatoryOnlyCases: Return cases for VD parameter where only mandatory sub-parameters are defined
+--[[ @isSubscribable: Check whether VD parameter is subscribable
+--! E.g. it's no possible to subscribe to 'vin' VD parameter
```
Done in 022433c

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -7,20 +7,17 @@ local runner = require('user_modules/script_runner')
-local apiLoader = require("modules/api_loader")
 local color = require("user_modules/consts").color
+local ah = require("user_modules/api/APIHelper")
+local tdg = require("user_modules/api/APITestDataGenerator")
```
Done in 022433c

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+    return out
+  end
+
+  local function getMandatoryChildren(pGraph, pId, pChildreIds)
```
Done in 022433c

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+  end
+
+  local function getMandatoryNeighbors(pGraph, pId, pParentIds)
+    local pIds = utils.cloneTable(pParentIds)
```
Done in 022433c

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+  tcs = createTestCases(ah.apiType.HMI, rpcType, m.rpcHMIMap[rpc],
+    m.isMandatory.ALL, m.isArray.YES, m.isVersion.ALL, {})
+  for _, tc in pairs(tcs) do
+    tc.graph[tc.paramId].valueTypeArray = boundValueTypeMap[testType]
```
Done in 022433c

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+      end
+      return false
+    end
+    local function getMandatoryValues(pId, pLevel, pOut)
```
Done in 022433c

dboltovskyi commented:
@jacobkeeler Please notice all the comments addressed

jacobkeeler commented:
@dboltovskyi Also as a note, several of the tests seem to be causing Core crashes as of now:
```
----------------------------------------------------------------------------------------------------
1:	ABORTED	./test_scripts/API/VehicleData/GetVehicleData/007_GetVD_app_version_is_greater_than_parameter_version.lua	
2:	ABORTED	./test_scripts/API/VehicleData/OnVehicleData/005_OnVD_App_is_not_subscribed.lua	
3:	ABORTED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/001_UnsubscribeVD_Success.lua	
4:	ABORTED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/003_UnsubscribeVD_disallowed_after_PTU.lua	
5:	ABORTED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/006_UnsubscribeVD_2nd_request_IGNORED.lua	
6:	ABORTED	./test_scripts/API/VehicleData/UnsubscribeVehicleData/007_UnsubscribeVD_for_2_apps.lua	
----------------------------------------------------------------------------------------------------
TOTAL: 6
PASSED: 0 
FAILED: 0 
ABORTED: 6 
SKIPPED: 0 
----------------------------------------------------------------------------------------------------
Execution time: 00:07:20
====================================================================================================
```

These crashes seem to be random, but they always happen during an `UnsubscribeVehicleData` step:
```
==============================
Start './test_scripts/API/VehicleData/UnsubscribeVehicleData/006_UnsubscribeVD_2nd_request_IGNORED.lua'
==============================
SDL build option REMOTE_CONTROL is unavailable.
Assume that SDL was built with REMOTE_CONTROL = ON
Default mobile device transport: TCP
Testing VD parameters are not restricted
Disabled VD parameter:	clusterModeStatus
=====================================================================================================
Description: ATF test script
Severity: Major
=====================================================================================================
--- Preconditions -----------------------------------------------------------------------------------
kill: (127953): No such process
[08:41:39,507] Clean_environment_and_update_preloaded_pt_file                                        [SUCCESS] (404 ms)
SDL pid 127989
tcp        0      0 127.0.0.1:8087          0.0.0.0:*               LISTEN     
SDL started
HMI initialized
HMI is ready
Mobile #1 connected
SDL allowed
[08:41:42,605] Start_SDL__HMI__connect_Mobile__start_Session                                         [SUCCESS] (1467 ms)
[08:41:44,093] Register_App                                                                          [SUCCESS] (2917 ms)
--- Test --------------------------------------------------------------------------------------------
--- VD parameter: accPedalPosition ------------------------------------------------------------------
[08:41:47,011] RPC_SubscribeVehicleData_SUCCESS                                                      [SUCCESS] (396 ms)
[08:41:47,407] RPC_1_UnsubscribeVehicleData_SUCCESS                                                  [SUCCESS] (403 ms)
[08:41:47,811] RPC_2_UnsubscribeVehicleData_IGNORED                                                  [SUCCESS] (1196 ms)
--- VD parameter: airbagStatus ----------------------------------------------------------------------
[08:41:49,008] RPC_SubscribeVehicleData_SUCCESS                                                      [SUCCESS] (401 ms)
SDL has unexpectedly crashed or stop responding!
Mobile #1 disconnected
[08:41:49,409] RPC_1_UnsubscribeVehicleData_SUCCESS                                                  [FAIL] (11196 ms)
 OnHashChange notification: Timeout: OnHashChange notification: Timeout expired
 Response to 7: Timeout: Response to 7: Timeout expired
SDL had already stopped
SCRIPT STATUS: ABORTED

```

Seeing a SIGSEGV message in the logs:
```
TRACE [08 Jan 2021 08:41:49,417][139724782024448][Commands] /home/jacobkeeler/sdl_core/src/components/application_manager/src/commands/command_request_impl.cc:798 commands::CommandRequestImpl::IsHMIResultSuccess: Enter
TRACE [08 Jan 2021 08:41:49,417][139724782024448][Commands] /home/jacobkeeler/sdl_core/src/components/application_manager/src/commands/command_request_impl.cc:798 commands::CommandRequestImpl::IsHMIResultSuccess: Exit
TRACE [08 Jan 2021 08:41:49,417][139724782024448][Commands] /home/jacobkeeler/sdl_core/src/components/application_manager/src/commands/command_request_impl.cc:813 commands::CommandRequestImpl::PrepareResultForMobileResponse: Exit
TRACE [08 Jan 2021 08:41:49,417][139724782024448][VehicleInfoPlugin] /home/jacobkeeler/sdl_core/src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/custom_vehicle_data_manager_impl.cc:89 CustomVehicleDataManagerImpl::CreateMobileMessageParams: Enter
TRACE [08 Jan 2021 08:41:49,417][139724782024448][VehicleInfoPlugin] /home/jacobkeeler/sdl_core/src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/custom_vehicle_data_manager_impl.cc:89 CustomVehicleDataManagerImpl::CreateMobileMessageParams: Exit
DEBUG [08 Jan 2021 08:41:49,417][139724782024448][Commands] /home/jacobkeeler/sdl_core/src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/commands/mobile/unsubscribe_vehicle_data_request.cc:316 commands::UnsubscribeVehicleDataRequest::UnsubscribePendingVehicleData: Unsubscribing from all pending VehicleData
DEBUG [08 Jan 2021 08:41:49,417][139724782024448][VehicleInfoPlugin] /home/jacobkeeler/sdl_core/src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/vehicle_info_app_extension.cc:67 VehicleInfoAppExtension::unsubscribeFromVehicleInfo: airbagStatus
DEBUG [08 Jan 2021 08:41:49,417][139724782024448][SDLMain] /home/jacobkeeler/sdl_core/src/appMain/life_cycle_impl.cc:240 {anonymous}::sig_handler: SIGSEGV signal has been caught
```

dboltovskyi commented:
@jacobkeeler Please notice the following:

1. Regarding the crashes, we found out they are not related to ATF tests, but to some system updates in Ubuntu.
We can't see them on an older version of Ubuntu 18 (~Sep/2020) or on Ubuntu 16 (latest).
A new issue needs to be created against `sdl_core` to address it.

2. Regarding the removed code
It's related to `isSkipped()` function and allows to have additional `LOWER_OUT_OF_BOUND` tests for `String` (and `Array`) parameters which have `minlength` (or `minsize`) = 0. This is related to the [2488#discussion_r553757334](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#discussion_r553757334).

dboltovskyi commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -7,25 +7,24 @@ local runner = require('user_modules/script_runner')
+  m.runner = {
+    Title = runner.Title,
+    Step = runner.Step
+  }
```
Done in 33ecefb

dboltovskyi commented:
@jacobkeeler Please notice all the questions have been addressed

jacobkeeler commented on user_modules/api/APITestDataGenerator.lua:
```diff
@@ -0,0 +1,309 @@
+  local length
+  if pTypeData.valueType == m.valueType.LOWER_IN_BOUND then
+    length = pTypeData.minlength
+    if not length or length == 0 then length = api.dataType.STRING.min end
```
```suggestion
    if not length then length = api.dataType.STRING.min end
```

dboltovskyi replied:
Done in 0ed0eed

jacobkeeler commented on user_modules/api/APITestDataGenerator.lua:
```diff
@@ -0,0 +1,309 @@
+  if pTypeData.array == true then
+    if arrayValueType == m.valueType.LOWER_IN_BOUND then
+      numOfItems = pTypeData.minsize
+      if not numOfItems then numOfItems = 0 end
```
```suggestion
      if not numOfItems then numOfItems = 1 end
```

jacobkeeler commented on user_modules/api/APITestDataGenerator.lua:
```diff
@@ -0,0 +1,309 @@
+      numOfItems = pTypeData.maxsize
+      if not numOfItems or numOfItems == 0 then numOfItems = 100 end
+    elseif arrayValueType == m.valueType.LOWER_OUT_OF_BOUND then
+      numOfItems = pTypeData.minsize
```
```suggestion
      numOfItems = pTypeData.minsize
      if not numOfItems then numOfItems = 1 end
      numOfItems = numOfItems - 1
```

jacobkeeler commented on test_scripts/API/VehicleData/common.lua:
```diff
@@ -636,4 +542,730 @@ function m.setAppVersion(pParamVersion, pOperator)
+  for _, tc in pairs(tcs) do
+    local function isSkipped()
+      local paramData = tc.graph[tc.paramId]
+      if (testType == m.testType.LOWER_OUT_OF_BOUND and (paramData.minsize == 0 or paramData.minsize == nil))
```
```suggestion
      if (testType == m.testType.LOWER_OUT_OF_BOUND and paramData.minsize == 0)
```

jacobkeeler reviewed (changes requested):
This should cover all of the changes that need to be made for the IN_BOUND/OUT_OF_BOUND tests

dboltovskyi commented:
@jacobkeeler Please notice all the comments addressed.

The following 3 suggested updates were not applied due to discussion: [2488#commitcomment-45860121](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#commitcomment-45860121)
 - [2488#discussion_r555306621](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#discussion_r555306621)
 - [2488#discussion_r555307822](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#discussion_r555307822)
 - [2488#discussion_r555311276](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2488#discussion_r555311276)

jacobkeeler commented:
@dboltovskyi After checking the actual RPC parser code, I do see that `minsize` does default to 0. Going off of memory, I was convinced that it was treated the same as `minlength`, but it doesn't appear that is the case, my apologies.

dboltovskyi commented:
@jacobkeeler Please notice descriptions have been updated in afb7faf and conflicts resolved in e490871

jacobkeeler reviewed (commented):
Some suggestions for clarity
